src/ServiceControl/Recoverability/Retrying/Infrastructure/ReturnToSenderDequeuer.cs

```diff
@@ -45,37 +46,37 @@ public ReturnToSenderDequeuer(TransportDefinition transportDefinition, ReturnToS
             if (shouldProcess(message))
             {
                 await returnToSender.HandleMessage(message, sender);
-                IncrementCounterOrProlongTimer();
+                await IncrementCounterOrProlongTimer();
             }
             else
             {
                 Log.WarnFormat("Rejecting message from staging queue as it's not part of a fully staged batch: {0}", message.MessageId);
             }
         }
 
-        void IncrementCounterOrProlongTimer()
+        Task IncrementCounterOrProlongTimer()
         {
             if (IsCounting)
             {
-                CountMessageAndStopIfReachedTarget();
-            }
-            else
-            {
-                Log.Debug("Resetting timer");
-                timer.Change(TimeSpan.FromSeconds(45), Timeout.InfiniteTimeSpan);
+                return CountMessageAndStopIfReachedTarget();
             }
+
+            Log.Debug("Resetting timer");
+            timer.Change(TimeSpan.FromSeconds(45), Timeout.InfiniteTimeSpan);
+            return TaskEx.CompletedTask;
         }
 
-        void CountMessageAndStopIfReachedTarget()
+        Task CountMessageAndStopIfReachedTarget()
         {
             var currentMessageCount = Interlocked.Increment(ref actualMessageCount);
             Log.DebugFormat("Handling message {0} of {1}", currentMessageCount, targetMessageCount);
             if (currentMessageCount >= targetMessageCount.GetValueOrDefault())
             {
                 Log.DebugFormat("Target count reached. Shutting down forwarder");
-                // NOTE: This needs to run on a different thread or a deadlock will happen trying to shut down the receiver
-                Task.Run(StopInternal);
```

Comment:
The intent of this line was to preserve the previous behavior. The intent of the previous code was from a caller perspective make this a fire&forget operation and thus offloading from the current thread that is entering and triggering the stop. These changes now remove that and StopInternal is executed on the thread entering StopInternal which can cause issues (if the original comments still holds true). I would vote to not make those changes except for the Ignore introduction

Reply:
if there is something causing StopInternal to cause issues, then StopInternal should take care of the `Task.Run` (Which it already does)?

Reply:
Right now it does not. It uses Task.Run to dispatch the TrySetResult of the TCS but then awaits that. The intent here is to offload the whole StopInternal call away from the caller entering the count check. Happy to explain synch if needed.